sop-extras/src/Data/SOP/NonEmpty.hs

```diff
@@ -0,0 +1,175 @@
+{-# LANGUAGE DataKinds                #-}
+{-# LANGUAGE DeriveTraversable        #-}
+{-# LANGUAGE FlexibleContexts         #-}
+{-# LANGUAGE GADTs                    #-}
+{-# LANGUAGE InstanceSigs             #-}
+{-# LANGUAGE PolyKinds                #-}
+{-# LANGUAGE ScopedTypeVariables      #-}
+{-# LANGUAGE StandaloneDeriving       #-}
+{-# LANGUAGE StandaloneKindSignatures #-}
+{-# LANGUAGE TypeApplications         #-}
+{-# LANGUAGE TypeOperators            #-}
+{-# LANGUAGE UndecidableInstances     #-}
+
+-- | Type-level non-empty lists
+module Data.SOP.NonEmpty (
+    NonEmpty (..)
+    -- * Proofs
+  , IsNonEmpty (..)
+  , ProofNonEmpty (..)
+  , checkIsNonEmpty
+    -- * Working with 'NonEmpty'
+  , nonEmptyFromList
+  , nonEmptyHead
+  , nonEmptyInit
+  , nonEmptyLast
+  , nonEmptyMapOne
+  , nonEmptyMapTwo
+  , nonEmptyStrictPrefixes
+  , nonEmptyToList
+  ) where
+
+import           Control.Applicative
+import qualified Data.Foldable as Foldable
+import           Data.Kind (Type)
+import           Data.Proxy
+import           Data.SOP.Sing
+
+{-------------------------------------------------------------------------------
+  NonEmpty
+-------------------------------------------------------------------------------}
+
+-- | Non-empty variation on 'AtMost'
+type NonEmpty :: [Type] -> Type -> Type
+data NonEmpty xs a where
```

Comment:
Similar comments as for `AtMost` apply, an IMO more fitting name would be `NonEmptyPrefix`.

Another observation: `NonEmpty xs a` is isomorphic to `Telescope (K a) (K a) xs`/`SimpleTelescope (K a) xs`, but I am not sure if that should inform our choice of name (`VerySimpleTelescope` doesn't sound very helpful).

Reply:
Phew that type :sweat_smile: 

In any case, I think we are in agreement about:
- `AtMost` -> `Prefix`
- `NonEmpty` -> `NonEmptyPrefix`

But I shall do (and we can discuss further) these renamings in a separate PR.